### PR TITLE
Updated estimated order amount on total change for blocks

### PR DESCRIPTION
**Pull request by dpanta94**
### All Submissions:

* [x] Does your code follow the [Extendables](https://extendomattic.wordpress.com/standardizations/) standards?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changes proposed in this Pull Request:
Listening for changes on the total amount for the checkout and/or the currency code. If that happens using the Components key attribute, we force a remount and render the new Amazon Pay button using the updated estimatedOrderAmount value.

### How to test the changes in this Pull Request:

1. Estimated order amount should get updated after the checkout total is changed in cart and checkout blocks

### Other information:

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### CODE REVIEW

File: src/payments-methods/express/payment-methods-express.js [hunk 1]
```diff
@@ -14,19 +14,48 @@ import { renderAmazonButton } from '../../renderAmazonButton';
...
+/**
+ * Returns the estimated order amount button attribute.
+ * @param {object} props
+ * @returns {object}|null Estimated order amount button attribute.
+ */
+const calculateEstimatedOrderAmount = ( props ) => {
+	const { billing } = props;
+	const { currency } = billing;
+
+	const stringCartTotal     = String( billing.cartTotal.value );
```
Comment (kalessil) [inline src/payments-methods/express/payment-methods-express.js:35]: What's the data type of `billing.cartTotal.value`?

File: src/payments-methods/express/payment-methods-express.js [hunk 2]
```diff
@@ -14,19 +14,48 @@ import { renderAmazonButton } from '../../renderAmazonButton';
...
+	const cartTotalLength     = stringCartTotal.length;
+	const decimals            = currency.minorUnit;
```
Comment (kalessil) [inline src/payments-methods/express/payment-methods-express.js:37]: Where `currency` is coming from?

File: src/payments-methods/express/payment-methods-express.js [hunk 3]
```diff
@@ -14,19 +14,48 @@ import { renderAmazonButton } from '../../renderAmazonButton';
...
+	const checkOutValue       = stringCartTotal.slice( 0, cartTotalLength - decimals ) + '.' + stringCartTotal.slice( cartTotalLength - decimals );
```
Comment (kalessil) [inline src/payments-methods/express/payment-methods-express.js:38]: Not sure what's happening here. Can you elaborate, please?

File: src/payments-methods/express/payment-methods-express.js [hunk 4]
```diff
@@ -14,19 +14,48 @@ import { renderAmazonButton } from '../../renderAmazonButton';
...
  * Returns the Components that will be used by Amazon Pay "Express".
  *
  * @param {object} props
  * @returns React Component
  */
 export const AmazonExpressContent = ( props ) => {
-	return <AmazonPayExpressBtn { ...props } />;
+	const estimatedOrderAmount = calculateEstimatedOrderAmount( props );
+
+	const [id, setId] = useState( estimatedOrderAmount ? estimatedOrderAmount.amount + estimatedOrderAmount.currencyCode : '0' );
```
Comment (kalessil) [inline src/payments-methods/express/payment-methods-express.js:52]: Could we do something like that, for better clarity?
```
const key = estimatedOrderAmount ? `${ estimatedOrderAmount.amount }${ estimatedOrderAmount.currencyCode }` : '0' 
```

File: src/payments-methods/express/payment-methods-express.js (hunk 2 above)
Comment (kalessil) [reply src/payments-methods/express/payment-methods-express.js:37]: Found!

File: src/payments-methods/express/payment-methods-express.js (hunk 1 above)
Comment (dpanta94) [reply src/payments-methods/express/payment-methods-express.js:35]: Its Number

File: src/payments-methods/express/payment-methods-express.js (hunk 2 above)
Comment (dpanta94) [reply src/payments-methods/express/payment-methods-express.js:37]: Is being set 2 lines above on the same method. Originally, some `props` are being fed into the `calculateEstimatedOrderAmount` function. currency is located in `props.billing.currency`

File: src/payments-methods/express/payment-methods-express.js (hunk 3 above)
Comment (dpanta94) [reply src/payments-methods/express/payment-methods-express.js]: @kalessil i added doc comments in every step of the function. Could you let me know if the functionality is clear now ?

File: src/payments-methods/express/payment-methods-express.js (hunk 4 above)
Comment (dpanta94) [reply src/payments-methods/express/payment-methods-express.js]: Yes of course, amended!

File: src/payments-methods/express/payment-methods-express.js [hunk 5]
```diff
@@ -14,19 +14,65 @@ import { renderAmazonButton } from '../../renderAmazonButton';
...
  *
  * @param {object} props
  * @returns React Component
  */
 export const AmazonExpressContent = ( props ) => {
-	return <AmazonPayExpressBtn { ...props } />;
+	const estimatedOrderAmount = calculateEstimatedOrderAmount( props );
+
+	const key = estimatedOrderAmount ? `${ estimatedOrderAmount.amount }${ estimatedOrderAmount.currencyCode }` : '0';
+
```
Comment (kalessil) [inline src/payments-methods/express/payment-methods-express.js:68]: We can omit the empty line here.

**Review by kalessil — APPROVED** [review]
LGTM!

File: src/payments-methods/express/payment-methods-express.js [hunk 6]
```diff
@@ -14,19 +14,65 @@ import { renderAmazonButton } from '../../renderAmazonButton';
...
+	const [id, setId] = useState( key );
+
+	useEffect( () => {
+		setId( estimatedOrderAmount ? estimatedOrderAmount.amount + estimatedOrderAmount.currencyCode : '0' );
```
Comment (kalessil) [inline src/payments-methods/express/payment-methods-express.js:70]: Can we use `key` here?

Comment (dpanta94) [reply src/payments-methods/express/payment-methods-express.js]: Yes, you are right. Much better

**Pull request merged**